multiraft/heartbeat_test.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2014 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Tobias Schottdorf (tobias.schottdorf@gmail.com)
+
+package multiraft
+
+import (
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/cockroachdb/cockroach/util/log"
+	"github.com/coreos/etcd/raft/raftpb"
+)
+
+func processEventsUntil(ch <-chan *interceptMessage, f func(*RaftMessageRequest) bool) {
+	for e := range ch {
+		e.ack <- struct{}{}
+		if f(e.args.(*RaftMessageRequest)) {
+			return
+		}
+	}
+}
+
+// validateHeartbeatCount receives messages from the given channel and compares
+// the observed number of heartbeat requests and responses.
+// Once the channel closes or enough heartbeat responses have been observed,
+// the function returns.
+func validateHeartbeatCount(ch <-chan *interceptMessage,
+	heartbeatCount, heartbeatResponseCount int, t *testing.T) {
+	// maps nodeID to heartbeats sent by this node minus heartbeat
+	// responses received by this node.
+	outInDelta := make(map[uint64]int)
+	cntH := 0
+	cntR := 0
+	defer func() {
+		for nodeID, delta := range outInDelta {
+			// If one of the checks below fails, this information will be useful.
+			log.Infof("node %v: %d heartbeats without response", nodeID, delta)
+		}
+		if cntH != heartbeatCount {
+			t.Errorf("unexpected total number of heartbeats: %d (expected %d)",
+				cntH, heartbeatCount)
+		}
+		if cntR != heartbeatResponseCount {
+			t.Errorf("unexpected total number of heartbeat responses: %d (expected %d)",
+				cntR, heartbeatResponseCount)
+		}
+	}()
+	for iMsg := range ch {
```

Comment:
I prefer to only use "range ch" when we expect the loop to terminate by the closing of the channel. It doesn't look like that's the case here, so I would use `for { iMsg := <-ch }`. This makes it more clear that the return at line 83 is the only exit (which means we could make it a break instead of a return and move the defer to the main body of the function).